### ISSUE TITLE
post save hook incorrect binding of this variable

### ISSUE THREAD

**Issue by nikoskalogridis** [Automattic/mongoose#2816]
In mongoose 3.8.x the `this` variable on the post save hook function equals to the document object:

```
MySchema.post('save', function () {
    console.log('entered post save hook', this);
});
```

```
entered post save hook
{ _id: 532572babfceadc93287cf92,
  __v: 99,
  created: Wed Mar 26 2014 08:30:13 GMT+0200 (EET),
  [...],
  id: '532572babfceadc93287cf92' }
```

but in version 4.x it is not:

```
entered post save hook
{ domain: null,
  _events: 
   { save: 
      [ [Function: notify],
        [Function: notify],
        [Function: notify],
        [Function: notify],
        [Function: notify],
        [Function: notify],
        [Function],
        [Function: notify],
        [Function: notify],
        [Function: notify],
        [Function: notify],
        [Function: notify],
        [Function: notify] ],
     isNew: 
      [ [Function: notify],
        [Function: notify],
        [Function: notify],
        [Function: notify],
        [Function: notify],
        [Function: notify],
        [Function: notify],
        [Function: notify],
        [Function: notify],
        [Function: notify],
        [Function: notify],
        [Function: notify] ] },
  _maxListeners: 0 }
```

**Comment by vkarpov15** [comment]
That's unexpected. Thanks for catching it.

**Comment by vkarpov15** [comment]
As a workaround, [post hooks also take the updated doc as the first parameter](http://mongoosejs.com/docs/middleware.html#post). That's the preferred way of getting the document in the post save hook.

**Comment by nikoskalogridis** [comment]
Yes, I have noticed that new feature. The problem is that I have to carefully review my code and make this change in all cases. So I will refrain from switching to 4.x until either that is fixed or I find some time to make the change to my code.

A notice on the migration doc should be made though in case other people use this

**Comment by vkarpov15** [comment]
Not a new feature, that's also the documented way to get access to the doc in the post save hook in 3.8 too. The reason why this slipped under the radar is that accessing `this` in a post hook was not officially supported behavior in 3.x (and so #1351 means that post hooks are getting the document's event emitter rather than the document), but we can support this convention going forward.

**Comment by asiapatrick** [comment]
I believe many modules required "this" to work.
e.g.
https://www.npmjs.com/package/mongoose-history

My site was down due to upgrade to mongoose v4.   When "this" could be access in "pre" and "post"?

**Comment by terebentina** [comment]
You too use `this` in the guide: http://mongoosejs.com/docs/middleware.html#notes

**Comment by cbankester** [comment]
Could we add this to the 4.0. upgrade guide in the Wiki? I just spent about 30 minutes too long debugging this issue.

**Comment by vkarpov15** [comment]
@terebentina that's for query middleware, and also doesn't rely on `this` being the returned doc. You should be able to access `this` properly in all `pre` middlewares, just 4.0.1 has an issue with synchronous document post hooks that's fixed in #2844
